### ISSUE TITLE
Add cluster leader election

### ISSUE THREAD

**Issue by woodsaj** [raintank/worldping-api#6]
<a href="https://github.com/woodsaj"><img src="https://avatars.githubusercontent.com/u/8415072?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [woodsaj](https://github.com/woodsaj)**
_Friday Jun 19, 2015 at 20:22 GMT_
_Originally opened as https://github.com/raintank/grafana/issues/228_

---

There are a few features within the code base that should only be run from one node at a time.
This requires having the nodes co-ordinate this role amongst themselves.

Raft seems to be the new hotness when it comes to these things, so we should use that.  Coreos' etcd package has an implementation of raft.
https://godoc.org/github.com/coreos/etcd/raft

**Comment by woodsaj** [comment]
<a href="https://github.com/Dieterbe"><img src="https://avatars.githubusercontent.com/u/20774?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Dieterbe](https://github.com/Dieterbe)**
_Tuesday Jun 23, 2015 at 00:06 GMT_

---

1) mind sharing a little bit what those features are?
2) can we get away with transactions on the database?
3) for alerting, i noticed you mentioned somewhere running only 1 job producer, but i thought we decided we actually _wanted_ to run multiple alert job producers for HA, because if jobs get consistently routed (by key), the consumers will drop jobs they've already processed anyway. this is a fairly simplistic method of HA.  if you're thinking of running only 1 producer, and it dies and restarts somewhere else then we also need to keep track of the last timestamp at which jobs were scheduled. in case it takes several seconds to restart a producer, the new producer should also process the missed ticks from the last few seconds.  (i actually like this approach, it seems more efficient, but also requires more operations/automation, perhaps we should postpone this improvement until we're at a point where multiple producers bring too much overhead?)

**Comment by woodsaj** [comment]
<a href="https://github.com/woodsaj"><img src="https://avatars.githubusercontent.com/u/8415072?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [woodsaj](https://github.com/woodsaj)**
_Tuesday Jun 23, 2015 at 15:10 GMT_

---

This is a long term goal to meet future scalability needs.

1) alerting scheduler and also collector session management.

2) yes, and that is likely what will be deployed first.  but it does not scale.  So long term we need a better solution.

3) also true, but as with 2, does not scale.  If we are running 10 instances of grafana, we don't want to have all 10 pushing the same messages into the queue.

**Comment by woodsaj** [comment]
<a href="https://github.com/Dieterbe"><img src="https://avatars.githubusercontent.com/u/20774?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Dieterbe](https://github.com/Dieterbe)**
_Tuesday Jun 23, 2015 at 19:59 GMT_

---

> This is a long term goal to meet future scalability needs.

I believe @nopzor1200 described the raft leader election as a high-prio item that was a must before we can launch.

I agree with your reasoning @woodsaj but we should make sure we're on the same page regarding urgency and timeline of this.
also, to make this viable I will need to make the alerting scheduler stateful (keeping track of last successfully processed timestamp, perhaps this could go into the raft log or in etcd, or in the database. will we have a HA transactional database?)

**Comment by woodsaj** [comment]
<a href="https://github.com/Dieterbe"><img src="https://avatars.githubusercontent.com/u/20774?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Dieterbe](https://github.com/Dieterbe)**
_Wednesday Jul 08, 2015 at 02:30 GMT_

---

Just saw a docker talk at pre gophercon party about libkv which provides a nice abstraction for leader election (supports etcd, consul and zk)

**Comment by woodsaj** [comment]
<a href="https://github.com/nopzor1200"><img src="https://avatars.githubusercontent.com/u/9953157?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [nopzor1200](https://github.com/nopzor1200)**
_Saturday Jul 18, 2015 at 05:23 GMT_

---

I originally misunderstood whether this was a high prio vs low prio item @woodsaj confirm it (raft or the like) is not something we need to worry about for now right?

**Comment by woodsaj** [comment]
<a href="https://github.com/woodsaj"><img src="https://avatars.githubusercontent.com/u/8415072?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [woodsaj](https://github.com/woodsaj)**
_Sunday Jul 19, 2015 at 13:24 GMT_

---

This is low prio.

**Comment by woodsaj** [comment]
<a href="https://github.com/Dieterbe"><img src="https://avatars.githubusercontent.com/u/20774?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Dieterbe](https://github.com/Dieterbe)**
_Friday Jul 31, 2015 at 15:54 GMT_

---

(interestingly, this ticket was in "to do" in  codetree. when i moved it to backlog it removed the backlog milestone. i guess cause it doesn't use milestones for backlog.
